doc/design/proposal-synchronous-hooks.md

```diff
@@ -0,0 +1,258 @@
+# Universal, synchronous and in-thread loader hooks
+
+For background and motivation, see https://github.com/nodejs/node/issues/52219. Prototype implementation is in https://github.com/joyeecheung/node/tree/sync-hooks.
+
+TL;DR: the top priority of this proposal is to allow sun-setting CJS loader monkey patching as non-breakingly as possible (which is why it needs to be synchronous and in-thread because that's how the CJS loader and `require()` works). Then it's API consistency with the existing `module.register()` off-thread hooks.
+
+Existing users of CJS loader monkey patching to look into:
+
+- pirates (used by Babel and nyc)
+- require-in-the-middle (used by many tracing agents)
+- yarn pnp
+- tsx
+- ts-node
+
+## API design
+
+A high-level overview of the API.
+
+```js
+const { addHooks, removeHooks } = require('module');
+
+function resolve(specifier, context, nextResolve) {
+  const resolved = nextResolve(specifier, context);
+  return { url: resolved.url.replaceAll(/foo/g, 'bar'); }
+}
+
+function load(url, context, nextLoad) {
+  const loaded = nextLoad(specifier, context);
+  return { source: loaded.source.toString().replaceAll(/foo/g, 'bar'); }
+}
+
+// id is a symbol
+const id = addHooks({ resolve, load, ... });
+removeHooks(id);
+```
+
+1. The names `addHooks` and `removeHooks` take inspiration from pirates. Can be changed to other better names.
+2. An alternative design to remove the hooks could be `addHooks(...).unhook()`, in this case `addHooks()` returns an object that could have other methods.
+   1. This may allow third-party hooks to query itself and avoid double-registering. Though this functionality is probably out of scope of the MVP.
+3. Another alternative design would be like what pirates offer: `const revert = addHooks(..); revert();`.
+4. It seems useful to allow the results returned by the hooks to be partial i.e. hooks don't have to clone the result returned by the next (default) hook to override it, instead they only need to return an object with properties that they wish to override. This can save the overhead of excessive clones.
+5. It seems `shortCircuit` is not really necessary if hooks can just choose to not call the next hook?
+
+## Hooks
+
+### `resolve`: from specifier to url
+
+```js
+/**
+ * @typedef {{
+ *   parentURL?: string,
+ *   conditions?: string[],
+ *   importAttributes?: object[]
+ * }} ModuleResolveContext
+ */
+/**
+ * @typedef {{
+ *   url?: string,
```

Comment:
That's from the point 4 above - if it's not overriden, it's considered "no change" to avoid excessive clones...

Reply:
I don't know about this. It's very easy to screw up your cases/branches and sometimes a hook ends up returning nothing. We have `shortCircuit` to warn people early if they've done this. I don't think we want to encourage partial returns, even if it has a slight performance benefit; explicitness is less bug-prone.

Reply:
I do agree with "explicitness is less bug-prone", although this would add another overhead of checking that every hook invocation returns everything needed. I hope piling all the overhead didn't end up being a motivation to hold on to the monkey patching simply because it's more performant..

Reply:
Perhaps @JakobJingleheimer can chime in. His experience was what motivated `shortCircuit`: basically, it was very easy to write hooks that didn’t behave as expected, because errors weren’t thrown when required properties were missing.

Doesn’t the proposed design mean that later hooks in the chain are always called, or that they’re always called if the response isn’t “complete”? Like in the simplest possible chain where it’s my `resolve` and then `defaultResolve`, if my hook doesn’t return the optional `format` property, in this “everything is optional” API then I’d think that `defaultResolve` would run and potentially define `format`.
- What happens if `defaultResolve` throws on such a file? I’ve already returned early without calling `next`, but essentially `next` gets called regardless; so for something like resolving `file.coffee` I’d think we would get an error.
- What if I don’t _want_ `defaultResolve` to fill in the blanks? Like if I really want to leave `format` undefined, and figure it out in `load`, rather than getting the `format` as defined by `defaultResolve` even though I didn’t call `next`?

Alternatively, could this proposed behavior be opt-in via a value like `shortCircuit`, such as `supplement: true`? So the default behavior is the same as the async API, where Node throws on a hook return that’s missing required properties; but if this incomplete object has `supplement: true` then we’ve opted into the behavior you’re proposing here where it continues the chain to fill in the missing properties. We could add `supplement: true` to the async API too for consistency.

Reply:
> What happens if defaultResolve throws on such a file? I’ve already returned early without calling next, but essentially next gets called regardless; so for something like resolving file.coffee I’d think we would get an error.

In that case the hook should return a format. In the current design, you return `coffee` as the format in the resolve hook, and in the load hook you intercept all modules with `format: 'coffee'`. That's how the mini TypeScript transpiler in the WIP is implemented for example. `defaultResolve` doesn't throw when you return an unknown format. It just expands the path, and carry the format down to the following steps.

> What if I don’t want defaultResolve to fill in the blanks? Like if I really want to leave format undefined, and figure it out in load, rather than getting the format as defined by defaultResolve even though I didn’t call next?

You can use a specific format that's known to the hook, like 'custom' or something and intercept any module with this special format in the load hook and figure it out there.

Reply:
> In that case the hook should return a format.

So then if I want to short circuit, I need to ensure that I return a value for every supported property, otherwise `defaultResolve` will run and possibly error? And then we can never add new properties on the return shape in the future, because existing hooks written to short circuit by returning a complete object suddenly wouldn’t be complete anymore?

I understand the desire to avoid cloning, but how much of a performance impact is it? Could we preserve the existing API design from the async hooks as the default mode and have a way to opt into the avoid-cloning behavior?

Reply:
> I need to ensure that I return a value for every supported property, otherwise defaultResolve will run and possibly error?

For resolve, only `url` is mandatory if the hook wishes to short-circuit; I am pretty sure in the old design you need to do the same if you want to short-circuit anyway? Otherwise if you don't let the default resolve run (because it's short-circuited) and you don't return an url, where is the loader supposed to get the url?

Reply:
> I am pretty sure in the old design url is mandatory if you want to short-circuit anyway? Otherwise if you don’t let the default resolve run (because it’s short-circuited) and you don’t return an url, where is the loader supposed to get the url?

In the async hooks, `url` is always mandatory (from `resolve`). If you short-circuit and don’t return `url`, it’ll throw an error (or it should, anyway, I haven’t tried). But for the async hooks, we can add new optional properties and they won’t affect existing hooks, like if we want to let `resolve` return an `importAttributes` property or something. I think in your proposed design, even new optional properties being added would be a breaking change in the sense that hooks that currently short circuit by returning full objects would no longer be short-circuiting because their “full” objects are no longer full, since there’s a new optional property they would need to be returning now to make the return value full.

Getting back to the bigger picture, though, personally I feel like the UX of the existing hooks and how they chain is pretty solid, as it was based on familiar patterns like Express middleware, and it’s been in use for several years now and no one’s complained about it. Early on we got feedback about accidentally short-circuiting, which is why we added the `shortCircuit` property to force authors to be explicit when they wanted that behavior rather than doing it by accident. I feel like there should be a good reason to deviate from this pattern, and performance is a good reason but I’d like to know how big of a benefit we’re talking about and if there’s some way it can be opted into, so that the familiar behavior is the default.

Reply:
> I think in your proposed design, even new optional properties being added would be a breaking change in the sense that hooks that currently short circuit by returning full objects would no longer be short-circuiting because their “full” objects are no longer full, since there’s a new optional property they would need to be returning now to make the return value full.

I am not sure what is being described here, it seems self-contradictary. If the property say `importAttributes`, is optional, then it's not mandatory, then short-circuiting without returning it is fine.

> and no one’s complained about it. 

Have you seen https://github.com/nodejs/loaders/pull/198#issuecomment-2160588439?

Reply:
> Have you seen [#198 (comment)](https://github.com/nodejs/loaders/pull/198#issuecomment-2160588439)?

I don’t really know what that means. I thought it was referring to your proposal. @jsumners-nr can you clarify?

Regardless, the current hooks API has been around for a few years now, and aside from the feedback years ago which led us to add `shortCircuit` we haven’t gotten any feedback that I’m aware of that its DX was confusing. I really think we should mimic that API as closely as possible, as it’s proven and users are familiar with it, and the consistency is good for anyone looking to write hooks that work with both APIs (and consistency is good in general, of course).

Reply:
> I don’t really know what that means. I thought it was referring to your proposal. @jsumners-nr can you clarify?

When I read through https://github.com/DataDog/import-in-the-middle/blob/a8da141be1e0fe802cea9bdb83d39a61c3aea7f7/hook.js#L194-L244 it is difficult to understand what `parentResolve` is and why it needs to be invoked. I think it would be easier if the hook merely worked with the thing it said it is interested in (i.e. the result of the provided specifier being loaded).

Reply:
> When I read through [DataDog/import-in-the-middle@`a8da141`/hook.js#L194-L244](https://github.com/DataDog/import-in-the-middle/blob/a8da141be1e0fe802cea9bdb83d39a61c3aea7f7/hook.js#L194-L244) it is difficult to understand what `parentResolve` is and why it needs to be invoked.

I’ve never seen the `parentResolve` name before. In our docs it’s always `next` as in [`nextResolve`](https://nodejs.org/api/module.html#resolvespecifier-context-nextresolve) or [`nextLoad`](https://nodejs.org/api/module.html#loadurl-context-nextload). This follows the familiar [middleware pattern popularized by Express](https://expressjs.com/en/guide/using-middleware.html), like `function (req, res, next) { ... }`.

Reply:
The name doesn't matter. The issue is having to reason through what all _may_ be happening prior to the work you _want_ to do. Basically, the simpler the API can be, the better.

Reply:
It could be simpler if we didn’t have chaining, but I think we need chaining: there are many use cases for a chain of two (user-provided hook plus the Node default hook): basically any time I want to get what Node would have returned and modify it, rather that reimplement what Node would have done and then modify it. And of course chains of more than two, like to transform TypeScript and also some template format like `.svelte`. I don’t know of a pattern more widely known and understood than the Express middleware pattern that we adopted for this.